### PR TITLE
Add additional profiler markers and update from begin/endsample to using(marker)

### PR DESCRIPTION
**Pull request by davidkline-ms**
This change refines the input system profiler markers by replacing begin/endsample calls to the preferred scoped using(marker.auto) syntax.

At the same time, additional markers were added that rounds out the suite of input markers.

As a part of this change, one HashSet was replaced with a Dictionary to eliminate some collection iterations. It has not been formally measured, though, anecdotally, it has been reported to have improved a partner application's performance.

### CODE REVIEW

File: Assets/MRTK/Core/Providers/Hands/ArticulatedHandDefinition.cs [hunk 1]
```diff
@@ -72,42 +72,44 @@ public bool IsInPointingPose
             }
         }
 
+        private static readonly ProfilerMarker UpdateHandJointsPerfMarker = new ProfilerMarker("[MRTK] ArticulatedHandDefinition.UpdateHandPoints");
```
Comment (keveleigh) [inline Assets/MRTK/Core/Providers/Hands/ArticulatedHandDefinition.cs:75]:
```suggestion
        private static readonly ProfilerMarker UpdateHandJointsPerfMarker = new ProfilerMarker("[MRTK] ArticulatedHandDefinition.UpdateHandJoints");
```

File: Assets/MRTK/Core/Providers/UnityInput/UnityJoystickManager.cs [hunk 2]
```diff
@@ -100,130 +101,124 @@ public override void Disable()
...
-            Profiler.EndSample(); // GetActiveControllers
-
-            return controllers;
+            using (GetActiveControllersPerfMarker.Auto())
+            {
+                IMixedRealityController[] controllers = ActiveControllers.Values.ToArray<IMixedRealityController>();
+                return controllers;
+            }
         }
+        private static readonly ProfilerMarker RefreshDevicesPerfMarker = new ProfilerMarker("[MRTK] UnityJoystickManager.Update");
```
Comment (keveleigh) [inline Assets/MRTK/Core/Providers/UnityInput/UnityJoystickManager.cs:115]: nit, to match the others
```suggestion

        private static readonly ProfilerMarker RefreshDevicesPerfMarker = new ProfilerMarker("[MRTK] UnityJoystickManager.Update");
```

File: Assets/MRTK/Providers/OpenVR/GenericOpenVRController.cs [hunk 3]
```diff
@@ -145,89 +145,91 @@ public GenericOpenVRController(TrackingState trackingState, Handedness controlle
             new MixedRealityInteractionMapping(12, "Spatial Grip", AxisType.SixDof, DeviceInputType.SpatialGrip),
         };
 
+        private static readonly ProfilerMarker UpdateControllerPerfMarker = new ProfilerMarker("[MRTK] GernericOpenVRController.UpdateController");
```
Comment (keveleigh) [inline Assets/MRTK/Providers/OpenVR/GenericOpenVRController.cs:148]:
```suggestion
        private static readonly ProfilerMarker UpdateControllerPerfMarker = new ProfilerMarker("[MRTK] GenericOpenVRController.UpdateController");
```

File: Assets/MRTK/Providers/OpenVR/GenericOpenVRController.cs [hunk 4]
```diff
@@ -145,89 +145,91 @@ public GenericOpenVRController(TrackingState trackingState, Handedness controlle
...
-            }
-
-            base.UpdateController();
 
-            Profiler.EndSample(); // UpdateController
+                base.UpdateController();
+            }
         }
 
+        private static readonly ProfilerMarker UpdateControllerDataPerfMarker = new ProfilerMarker("[MRTK] GernericOpenVRController.UpdateControllerData");
```
Comment (keveleigh) [inline Assets/MRTK/Providers/OpenVR/GenericOpenVRController.cs:174]:
```suggestion
        private static readonly ProfilerMarker UpdateControllerDataPerfMarker = new ProfilerMarker("[MRTK] GenericOpenVRController.UpdateControllerData");
```

File: Assets/MRTK/Providers/WindowsMixedReality/XR2018/Controllers/BaseWindowsMixedRealitySource.cs [hunk 5]
```diff
@@ -50,295 +50,301 @@ protected BaseWindowsMixedRealitySource(TrackingState trackingState, Handedness
 
         #region Update data functions
 
+        private static readonly ProfilerMarker UpdateControllerPerfMarker = new ProfilerMarker("[MRTK] BaseWindowsMixedRealitySource.UpdateContoller");
```
Comment (keveleigh) [inline Assets/MRTK/Providers/WindowsMixedReality/XR2018/Controllers/BaseWindowsMixedRealitySource.cs:53]:
```suggestion
        private static readonly ProfilerMarker UpdateControllerPerfMarker = new ProfilerMarker("[MRTK] BaseWindowsMixedRealitySource.UpdateController");
```

File: Assets/MRTK/Providers/WindowsMixedReality/XR2018/Controllers/WindowsMixedRealityArticulatedHand.cs [hunk 6]
```diff
@@ -90,88 +90,90 @@ private SpatialInteractionManager SpatialInteractionManager
...
+                            handDefinition?.UpdateCurrentIndexPose(Interactions[i]);
+                            break;
+                    }
                 }
             }
-
-            Profiler.EndSample(); // UpdateController
         }
 
+        private static readonly ProfilerMarker UpdateHandDataPerfMarker = new ProfilerMarker("[MRTK] WindowsMixedRealityArticulatedHand.UpdateHandData");
```
Comment (keveleigh) [inline Assets/MRTK/Providers/WindowsMixedReality/XR2018/Controllers/WindowsMixedRealityArticulatedHand.cs:118]: I _think_ this is only used in this context? To prevent a potential "field not used" warning.

```suggestion
#if WINDOWS_UWP || DOTNETWINRT_PRESENT
        private static readonly ProfilerMarker UpdateHandDataPerfMarker = new ProfilerMarker("[MRTK] WindowsMixedRealityArticulatedHand.UpdateHandData");
#endif // WINDOWS_UWP || DOTNETWINRT_PRESENT
```

File: Assets/MRTK/SDK/Features/UX/Scripts/Pointers/DefaultPointerMediator.cs [hunk 7]
```diff
@@ -32,189 +33,214 @@ public DefaultPointerMediator(IPointerPreferences pointerPrefs)
...
-                    ApplyPointerBehavior(pointer, pointerPreferences.GetPointerBehavior(pointer));
+                    foreach (IMixedRealityPointer pointer in allPointers)
+                    {
+                        ApplyPointerBehavior(pointer, pointerPreferences.GetPointerBehavior(pointer));
+                    }
                 }
             }
         }
 
+        private static readonly ProfilerMarker ApplyCustomPointerBehaviorPerfMarker = new ProfilerMarker("[MRTK] DefaultPointerMediator.ApplyCustomPointerBehavior");
```
Comment (keveleigh) [inline Assets/MRTK/SDK/Features/UX/Scripts/Pointers/DefaultPointerMediator.cs:224]: To match the method name:

```suggestion
        private static readonly ProfilerMarker ApplyPointerBehaviorPerfMarker = new ProfilerMarker("[MRTK] DefaultPointerMediator.ApplyPointerBehavior");
```

File: Assets/MRTK/SDK/Features/UX/Scripts/Pointers/DefaultPointerMediator.cs [hunk 8]
```diff
@@ -32,189 +33,214 @@ public DefaultPointerMediator(IPointerPreferences pointerPrefs)
...
+
         private void ApplyPointerBehavior(IMixedRealityPointer pointer, PointerBehavior behavior)
         {
-            if (behavior == PointerBehavior.Default)
+            using (ApplyCustomPointerBehaviorPerfMarker.Auto())
```
Comment (keveleigh) [inline Assets/MRTK/SDK/Features/UX/Scripts/Pointers/DefaultPointerMediator.cs:228]:
```suggestion
            using (ApplyPointerBehaviorPerfMarker.Auto())
```

File: Assets/MRTK/Services/InputSystem/FocusProvider.cs [hunk 9]
```diff
@@ -45,7 +45,7 @@ public class FocusProvider : BaseCoreSystem,
             focusIndividualCompoundCollider = profile.FocusIndividualCompoundCollider;
         }
 
-        private readonly HashSet<PointerData> pointers = new HashSet<PointerData>();
+        private readonly Dictionary<uint, PointerData> pointers = new Dictionary<uint, PointerData>();
```
Comment (keveleigh) [inline Assets/MRTK/Services/InputSystem/FocusProvider.cs:48]: Was this change intended for this PR too?
If so, do you happen to have perf measurements before / after? Looking at the code, I'd expect this to be better, but was wondering to be sure.

File: Assets/MRTK/Core/Providers/UnityInput/UnityJoystickManager.cs (hunk 2 above)
Comment (davidkline-ms) [reply Assets/MRTK/Core/Providers/UnityInput/UnityJoystickManager.cs:115]: There's another mistake in this marker... The string has the wrong method name.

File: Assets/MRTK/Services/InputSystem/FocusProvider.cs (hunk 9 above)
Comment (davidkline-ms) [reply Assets/MRTK/Services/InputSystem/FocusProvider.cs:48]: It was, but i forgot to add it to the description. :/

i dont have before / after. anecdotally, the partner we tested this change for states it is an improvement.

File: Assets/MRTK/Core/Providers/UnityInput/UnityTouchDeviceManager.cs [hunk 10]
```diff
@@ -54,46 +54,47 @@ public class UnityTouchDeviceManager : BaseInputDeviceManager
 
         private List<UnityTouchController> touchesToRemove = new List<UnityTouchController>();
 
+        private static readonly ProfilerMarker UpdatePerfMarker = new ProfilerMarker("[MRTK] UnityTouchManager.Update");
```
Comment (wiwei) [inline Assets/MRTK/Core/Providers/UnityInput/UnityTouchDeviceManager.cs:57]:
```suggestion
        private static readonly ProfilerMarker UpdatePerfMarker = new ProfilerMarker("[MRTK] UnityTouchDeviceManager.Update");
```

File: Assets/MRTK/Core/Providers/UnityInput/UnityTouchDeviceManager.cs [hunk 11]
```diff
@@ -117,88 +118,89 @@ public override void Disable()
             ActiveTouches.Clear();
         }
 
+        private static readonly ProfilerMarker AddTouchControllerPerfMarker = new ProfilerMarker("[MRTK] UnityTouchManager.AddTouchController");
```
Comment (wiwei) [inline Assets/MRTK/Core/Providers/UnityInput/UnityTouchDeviceManager.cs:121]:
```suggestion
        private static readonly ProfilerMarker AddTouchControllerPerfMarker = new ProfilerMarker("[MRTK] UnityTouchDeviceManager.AddTouchController");
```

File: Assets/MRTK/Core/Providers/UnityInput/UnityTouchDeviceManager.cs [hunk 12]
```diff
@@ -117,88 +118,89 @@ public override void Disable()
...
-            controller.StartTouch();
+                Service?.RaiseSourceDetected(controller.InputSource, controller);
 
-            Profiler.EndSample(); // AddTouchController
+                controller.TouchData = touch;
+                controller.StartTouch();
+            }
         }
 
+        private static readonly ProfilerMarker UpdateTouchDataPerfMarker = new ProfilerMarker("[MRTK] UnityTouchManager.UpdateTouchData");
```
Comment (wiwei) [inline Assets/MRTK/Core/Providers/UnityInput/UnityTouchDeviceManager.cs:162]:
```suggestion
        private static readonly ProfilerMarker UpdateTouchDataPerfMarker = new ProfilerMarker("[MRTK] UnityTouchDeviceManager.UpdateTouchData");
```

File: Assets/MRTK/Core/Providers/UnityInput/UnityTouchDeviceManager.cs [hunk 13]
```diff
@@ -117,88 +118,89 @@ public override void Disable()
...
 
-            Profiler.EndSample(); // UpdateTouchData
+                controller.TouchData = touch;
+                var pointer = (IMixedRealityTouchPointer)controller.InputSource.Pointers[0];
+                controller.ScreenPointRay = pointer.TouchRay = ray;
+                controller.Update();
+            }
         }
 
+        private static readonly ProfilerMarker RemoveTouchControllerPerfMarker = new ProfilerMarker("[MRTK] UnityTouchManager.RemoveTouchController");
```
Comment (wiwei) [inline Assets/MRTK/Core/Providers/UnityInput/UnityTouchDeviceManager.cs:182]:
```suggestion
        private static readonly ProfilerMarker RemoveTouchControllerPerfMarker = new ProfilerMarker("[MRTK] UnityTouchDeviceManager.RemoveTouchController");
```

File: Assets/MRTK/Core/Services/BaseDataProviderAccessCoreSystem.cs [hunk 14]
```diff
@@ -41,34 +41,36 @@ public override void Enable()
             }
         }
 
+        private static readonly ProfilerMarker UpdatePerfMarker = new ProfilerMarker("[MRTK] BaseDataProviderAccessCoreSystem.Update");
```
Comment (wiwei) [inline Assets/MRTK/Core/Services/BaseDataProviderAccessCoreSystem.cs:44]: It feels like just knowing that the base class is taking time isn't actually super helpful - if there were some way to also propagate the actual system that might make these markers more helpful here.

File: Assets/MRTK/Providers/WindowsMixedReality/XR2018/Controllers/WindowsMixedRealityController.cs [hunk 15]
```diff
@@ -66,184 +66,188 @@ public WindowsMixedRealityController(TrackingState trackingState, Handedness con
...
         /// <summary>
         /// Update the controller data from the provided platform state.
         /// </summary>
         /// <param name="interactionSourceState">The InteractionSourceState retrieved from the platform</param>
         public override void UpdateController(InteractionSourceState interactionSourceState)
         {
-            if (!Enabled) { return; }
-
-            Profiler.BeginSample("[MRTK] WindowsMixedRealityController.UpdateContoller");
+            using (UpdateControllerPerfMarker.Auto())
```
Comment (wiwei) [inline Assets/MRTK/Providers/WindowsMixedReality/XR2018/Controllers/WindowsMixedRealityController.cs:77]: Should this happen after the enabled check?

Comment (wiwei) [reply Assets/MRTK/Providers/WindowsMixedReality/XR2018/Controllers/WindowsMixedRealityController.cs:77]: Otherwise it's possible for us to get a lot of "quick" events which would make things seem fast

File: Assets/MRTK/Providers/WindowsMixedReality/XRSDK/Controllers/WindowsMixedRealityXRSDKArticulatedHand.cs [hunk 16]
```diff
@@ -66,38 +66,41 @@ public WindowsMixedRealityXRSDKArticulatedHand(TrackingState trackingState, Hand
...
 
+        private static readonly ProfilerMarker UpdateControllerPerfMarker = new ProfilerMarker("[MRTK] WindowsMixedRealityXRSDKArticulatedHand.UpdateController");
+
         /// <inheritdoc />
         public override void UpdateController(InputDevice inputDevice)
         {
-            if (!Enabled) { return; }
-
-            Profiler.BeginSample("[MRTK] WindowsMixedRealityXRSDKArticulatdHand.UpdateController");
+            using (UpdateControllerPerfMarker.Auto())
```
Comment (wiwei) [inline Assets/MRTK/Providers/WindowsMixedReality/XRSDK/Controllers/WindowsMixedRealityXRSDKArticulatedHand.cs:74]: Same comment on the enabled thing.

File: Assets/MRTK/SDK/Features/UX/Scripts/Pointers/BaseControllerPointer.cs [hunk 17]
```diff
@@ -452,77 +463,92 @@ public override int GetHashCode()
...
 
         #region IMixedRealityInputHandler Implementation
 
+        private static readonly ProfilerMarker OnInputUpPerfMarker = new ProfilerMarker("[MRTK] BaseControllerPointer.OnInputUp");
+
         /// <inheritdoc />
         public override void OnInputUp(InputEventData eventData)
         {
-            if(!IsInteractionEnabled) { return; }
+            using (OnInputUpPerfMarker.Auto())
```
Comment (wiwei) [inline Assets/MRTK/SDK/Features/UX/Scripts/Pointers/BaseControllerPointer.cs:501]: Similar Q in this file around IsInteractionEnabled checks (i.e. is it worth doing it if it's a no-op)

File: Assets/MRTK/Services/InputSystem/FocusProvider.cs (hunk 9 above)
Comment (davidkline-ms) [reply Assets/MRTK/Services/InputSystem/FocusProvider.cs:48]: added comment to description

File: Assets/MRTK/Core/Providers/UnityInput/UnityTouchDeviceManager.cs (hunk 11 above)
Comment (davidkline-ms) [reply Assets/MRTK/Core/Providers/UnityInput/UnityTouchDeviceManager.cs:121]: fixed.

File: Assets/MRTK/Core/Providers/UnityInput/UnityTouchDeviceManager.cs (hunk 12 above)
Comment (davidkline-ms) [reply Assets/MRTK/Core/Providers/UnityInput/UnityTouchDeviceManager.cs:162]: fixed

File: Assets/MRTK/Core/Providers/UnityInput/UnityTouchDeviceManager.cs (hunk 13 above)
Comment (davidkline-ms) [reply Assets/MRTK/Core/Providers/UnityInput/UnityTouchDeviceManager.cs:182]: fixed

File: Assets/MRTK/Core/Services/BaseDataProviderAccessCoreSystem.cs (hunk 14 above)
Comment (davidkline-ms) [reply Assets/MRTK/Core/Services/BaseDataProviderAccessCoreSystem.cs:44]: i can look into this. in many cases, the derived class will also appear in the trace as they typically call the base as well as their own custom code.

File: Assets/MRTK/Providers/WindowsMixedReality/XR2018/Controllers/WindowsMixedRealityController.cs (hunk 15 above)
Comment (davidkline-ms) [reply Assets/MRTK/Providers/WindowsMixedReality/XR2018/Controllers/WindowsMixedRealityController.cs:77]: that's a good question. by showing these quick-out events, we see that there is a method call adding _some_ overhead and that the stack is consistent between these and more expensive calls.

Adding the markers after the enabled check _would_ make it easier to see the impact by eliminating the 0ms events.

I'll consider and modify in a future PR. I will need to scrub the complete tree to find all instances of this type of tracing,

File: Assets/MRTK/Core/Services/BaseDataProviderAccessCoreSystem.cs (hunk 14 above)
Comment (keveleigh) [reply Assets/MRTK/Core/Services/BaseDataProviderAccessCoreSystem.cs:44]: I think this should should be in the scope of the derived class' markers, this will show up in the profiler's hierarchy under that marker and we'll be able to see the split of time for each system

File: Assets/MRTK/Core/Providers/UnityInput/UnityTouchDeviceManager.cs (hunk 10 above)
Comment (davidkline-ms) [reply Assets/MRTK/Core/Providers/UnityInput/UnityTouchDeviceManager.cs]: fixed

File: Assets/MRTK/Core/Services/BaseDataProviderAccessCoreSystem.cs (hunk 14 above)
Comment (davidkline-ms) [reply Assets/MRTK/Core/Services/BaseDataProviderAccessCoreSystem.cs:44]: @keveleigh, are you proposing adding implementations that are only markers and calls to the base class for each derived class? or are you indicating that the trace context should make it clear with the markers as they currently are?

File: Assets/MRTK/Providers/WindowsMixedReality/XR2018/Controllers/WindowsMixedRealityController.cs (hunk 15 above)
Comment (davidkline-ms) [reply Assets/MRTK/Providers/WindowsMixedReality/XR2018/Controllers/WindowsMixedRealityController.cs:77]: https://github.com/microsoft/MixedRealityToolkit-Unity/issues/7647

Comment (wiwei) [reply Assets/MRTK/Providers/WindowsMixedReality/XR2018/Controllers/WindowsMixedRealityController.cs:77]: FWIW I think it would be better to take the changes now. We're already in the code and I highlighted the cases that I saw (I don't THINK there are a ton more)

Comment (davidkline-ms) [reply Assets/MRTK/Providers/WindowsMixedReality/XR2018/Controllers/WindowsMixedRealityController.cs:77]: that's fair

File: Assets/MRTK/Core/Services/BaseDataProviderAccessCoreSystem.cs (hunk 14 above)
Comment (davidkline-ms) [reply Assets/MRTK/Core/Services/BaseDataProviderAccessCoreSystem.cs:44]: Per discussion, leaving this as-is and will evaluate if further refinements are needed after spending time using this change.
